### PR TITLE
Trend micro vision one

### PR DESCRIPTION
**Pull request by shaqnawe**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Updated incoming mapper attributes. Updated pytmv1 package to 0.8.0 to support 6 new custom script actions. Fixed an issue with endpoint and email activity data count actions, where query_op was not being parsed correctly. This integration will use the dockerfiles docker image here: [updated pytmv1 to 0.8.0 #23882](https://github.com/demisto/dockerfiles/pull/23882)

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3.yml [hunk 1]
```diff
@@ -3,1199 +3,1293 @@ commonfields:
...
+        - contextPath: VisionOne.Update_Status.code
+          description: HTTP status code of updating workbench alert status.
+          type: string
+        - contextPath: VisionOne.Update_Status.message
+          description: Message notifying user that the alert status has been updated to user defined status.
+          type: string
+      description: Updates the status of a workbench alert.
+      name: trendmicro-visionone-update-status
+    - arguments:
+        - description: 'List of object(s) made up of `filename`, `endpoint_name` or `agent_guid` and optional `description` and optional `parameter`. e.g. [{"filename":"test.ps1","endpoint_name":"test-endpoint1","description":"Run custom script","parameter":"some-string"}].'
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3.yml:1194]:
```suggestion
        - description: 'A Json list of object(s) made up of `filename`, `endpoint_name` or `agent_guid` and optional `description` and optional `parameter`. e.g. [{"filename":"test.ps1","endpoint_name":"test-endpoint1","description":"Run custom script","parameter":"some-string"}].'
```

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3.yml [hunk 2]
```diff
@@ -3,1199 +3,1293 @@ commonfields:
...
+        - contextPath: VisionOne.Run_Custom_Script.status
+          description: Status of running custom script.
+          type: number
+        - contextPath: VisionOne.Run_Custom_Script.task_id
+          description: Task ID generated after running custom script.
+          type: string
+      description: Runs a custom script on the specified endpoint or agentGuid.
+      name: trendmicro-visionone-run-custom-script
+    - arguments:
+        - description: 'Filter for retrieving a subset of the custom scripts list. e.g. {"fileName":"test-script.ps1","fileType":"powershell"}. If you want a list of all custom scripts available, run action with fields=`{}`.'
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3.yml:1208]:
```suggestion
        - description: 'A dictionary object with categories to filter for retrieving a subset of the custom scripts list. e.g. {"fileName":"test-script.ps1","fileType":"powershell"}. If you want a list of all custom scripts available, run action with fields=`{}`.'
```

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3.yml [hunk 3]
```diff
@@ -3,1199 +3,1293 @@ commonfields:
...
+          name: fields
+          required: true
+        - description: The query operator used to join fields to filter query. e.g. query_op="or".
+          name: query_op
+      outputs:
+        - contextPath: VisionOne.Get_Custom_Script_List.id
+          description: The id for custom script.
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3.yml:1215]:
```suggestion
          description: The ID for custom script.
```

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py [hunk 4]
```diff
@@ -1404,3 +1409,165 @@ def test_update_status(mocker):
...
+def run_custom_script_mock_response(*args, **kwargs):
+    with open("./test_data/run_custom_script.json") as f:
+        return_value: list[dict[str, Any]] = json.load(f)
+    return MultiResult(
+        result_code=ResultCode.SUCCESS,
+        response=MultiResp(items=[MsData(**data) for data in return_value]),
+    )
+
+
+def test_run_custom_script(mocker):
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1424]: Please add a given-when-then documentation.

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py [hunk 5]
```diff
@@ -1404,3 +1409,165 @@ def test_update_status(mocker):
...
+                    file_name="string",
+                    file_type=FileType.POWERSHELL,
+                    description="Script to delete duplicate values",
+                ),
+            ]
+        ),
+    )
+
+
+def test_get_custom_script_list(mocker):
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1471]: Please add a given-when-then documentation.

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py [hunk 6]
```diff
@@ -1404,3 +1409,165 @@ def test_update_status(mocker):
...
+
+# Mock function to download custom script
+def download_custom_script_mock_response(*args, **kwargs):
+    return Result(
+        result_code=ResultCode.SUCCESS,
+        response=TextResp(text="#!/bin/sh echo 'Download script'"),
+    )
+
+
+def test_download_custom_script(mocker):
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1491]: Please add a given-when-then documentation.

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py [hunk 7]
```diff
@@ -1404,3 +1409,165 @@ def test_update_status(mocker):
...
+def delete_custom_script_mock_response(*args, **kwargs):
+    with open("./test_data/delete_custom_script.json") as f:
+        return_value: dict[str, str] = json.load(f)
+    return Result(
+        result_code=ResultCode.SUCCESS,
+        response=NoContentResp(**return_value),
+    )
+
+
+def test_delete_custom_script(mocker):
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1511]: Please add a given-when-then documentation.

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py [hunk 8]
```diff
@@ -1404,3 +1409,165 @@ def test_update_status(mocker):
...
+def add_custom_script_mock_response(*args, **kwargs):
+    with open("./test_data/add_custom_script.json") as f:
+        return_value: dict[str, str] = json.load(f)
+    return Result(
+        result_code=ResultCode.SUCCESS,
+        response=AddCustomScriptResp(**return_value),
+    )
+
+
+def test_add_custom_script(mocker):
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1531]: Please add a given-when-then documentation.

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py [hunk 9]
```diff
@@ -1404,3 +1409,165 @@ def test_update_status(mocker):
...
+def update_custom_script_mock_response(*args, **kwargs):
+    with open("./test_data/update_custom_script.json") as f:
+        return_value: dict[str, str] = json.load(f)
+    return Result(
+        result_code=ResultCode.SUCCESS,
+        response=NoContentResp(**return_value),
+    )
+
+
+def test_update_custom_script(mocker):
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1558]: Please add a given-when-then documentation.

File: Packs/TrendMicroVisionOne/ReleaseNotes/4_1_0.md [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+##### Trend Micro Vision One Event Collector
+
+##### Trend Micro Vision One V3.
+
+- Added action to run a custom script on an endpoint.
+- Added action to get a list of available custom scripts in XDR Portal.
+- Added action to add a custom script to XDR Portal.
+- Added action to update a custom script in XDR Portal.
+- Added action to download a custom script from XDR Portal.
+- Added action to delete a custom script from XDR Portal.
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/ReleaseNotes/4_1_0.md:13]:
```suggestion
- Added new commands:
    - command 1
    - command 2
```
Please use this template, replace command # with the name of the command as mentioned in the yml.

**Review by YuvHayun — CHANGES_REQUESTED** [review]
Hey @shaqnawe thx a lot for your contribution, looks very good!
I left small comments.
@melamedbn let me know when your review is done.

File: Packs/TrendMicroVisionOne/ReleaseNotes/4_1_0.md (hunk 10 above)
Comment (shaqnawe) [reply Packs/TrendMicroVisionOne/ReleaseNotes/4_1_0.md]: Done.

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py (hunk 9 above)
Comment (shaqnawe) [reply Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1562]: Done

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py (hunk 4 above)
Comment (shaqnawe) [reply Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1425]: Done.

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py (hunk 5 above)
Comment (shaqnawe) [reply Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1471]: Done.

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py (hunk 6 above)
Comment (shaqnawe) [reply Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1492]: Done.

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py (hunk 7 above)
Comment (shaqnawe) [reply Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1513]: Done.

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py (hunk 8 above)
Comment (shaqnawe) [reply Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1534]: Done.

**Comment by shaqnawe** [comment]
@YuvHayun I have made the recommended changes, please let me know if anything else is needed. Appreciate your assistance on this!

File: Packs/TrendMicroVisionOne/ReleaseNotes/4_1_0.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
+
+#### Integrations
+
+##### Trend Micro Vision One Event Collector
+
+##### Trend Micro Vision One V3.
+
+- Added new custom script commands:
+  - trendmicro-visionone-run-custom-script runs a custom script on an endpoint.
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/ReleaseNotes/4_1_0.md:9]:
```suggestion
  - **trendmicro-visionone-run-custom-script** runs a custom script on an endpoint.
```
Let's do so for all commands.

**Review by YuvHayun — CHANGES_REQUESTED** [review]
Hey @shaqnawe, I see you're done with most changes.
I can't find the given-when-then documentation you mentioned you added to the tests.
Also, let me know how's it going and when do you want to merge the docker pr?

**Comment by shaqnawe** [comment]
@melamedbn Here's a link to the demo video for incoming classifier mapping. https://trendmicro-my.sharepoint.com/:v:/p/shakti_shah/EQS5f-P4s9BIuqSBDmsritEBcX0WJCxNXHne1OXneC0z8Q
Please let me know if you have any questions. Thank you!

**Comment by shaqnawe** [comment]
@melamedbn here's a screenshot of what the Vision One tab looks like in the UI.
Screenshot 2024-02-05 at 2 49 13 PM

File: Packs/TrendMicroVisionOne/ReleaseNotes/4_1_0.md (hunk 11 above)
Comment (shaqnawe) [reply Packs/TrendMicroVisionOne/ReleaseNotes/4_1_0.md:9]: done.

**Comment by shaqnawe** [comment]
> Hey @shaqnawe, I see you're done with most changes. I can't find the given-when-then documentation you mentioned you added to the tests. Also, let me know how's it going and when do you want to merge the docker pr?

@YuvHayun can you give an example of the given-when-then documentation?

**Comment by YuvHayun** [comment]
@shaqnawe no problem, I'll merge it.
Regarding the given-when-then.. you can see some examples in `crowdstrikefalcon_test.py`
It's just a docstring template.
Let me know when all errors are fixed and you added the docker & documentation.

**Review by melamedbn — APPROVED** [review]

File: Packs/TrendMicroVisionOne/ReleaseNotes/4_1_0.md [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+##### Trend Micro Vision One V3.
+
+- Added new custom script commands:
+  - **trendmicro-visionone-run-custom-script** runs a custom script on an endpoint.
+  - **trendmicro-visionone-get-custom-script-list** fetches a list of available custom scripts in XDR Portal.
+  - **trendmicro-visionone-add-custom-script** adds a custom script to XDR Portal.
+  - **trendmicro-visionone-update-custom-script** updates a custom script in XDR Portal.
+  - **trendmicro-visionone-download-custom-script** downloads a custom script from XDR Portal.
+  - **trendmicro-visionone-delete-custom-script** deletes a custom script from XDR Portal.
+- Updated the Docker image to: *devtesting/pytmv1:0.8.0.1706812855*.
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/ReleaseNotes/4_1_0.md:15]:
```suggestion
- Updated the Docker image to: *demisto/pytmv1:0.8.0.87051*.
```

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py [hunk 13]
```diff
@@ -1404,3 +1409,169 @@ def test_update_status(mocker):
     assert result.outputs["code"] == 204
     assert isinstance(result.outputs["message"], str)
     assert result.outputs["Workbench_Id"] == "WB-20837-20220418-00000"
+
+
+# Mock function for run custom script
+def run_custom_script_mock_response(*args, **kwargs):
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3_test.py:1415]: As mentioned above, you can find a good example for given-when-then doc strings in crowdstrikeflacon_test.py file.

**Review by YuvHayun — CHANGES_REQUESTED** [review]
Minor comments.

**Comment by YuvHayun** [comment]
Hey @shaqnawe, how are you doing with the fixes?

**Comment by shaqnawe** [comment]
> Hey @shaqnawe, how are you doing with the fixes?

@YuvHayun Still working on the change on our side. All the changes you've recommended have been added.

**Comment by YuvHayun** [comment]
@shaqnawe no problem, let me know when it's ready for re-review.

**Comment by shaqnawe** [comment]
@YuvHayun I have made the necessary changes, please re-review. Thank you!

**Comment by shaqnawe** [comment]
@maimorag  @YuvHayun not ready for re-review yet.

**Comment by shaqnawe** [comment]
@YuvHayun necessary changes have been made and PR is ready for review. Pytmv1 package has been updated to 0.8.1 and a PR for demisto-docker has been opened for updated dockerimage. Trend pytmv1 #24475 --> https://github.com/demisto/dockerfiles/pull/24475
I will update the release notes and yml once the new image is available on dockerhub/demisto.

**Comment by YuvHayun** [comment]
Hey @shaqnawe, np.
I'll let you know when the docker pr is merged and I'll review everything once the new docker is updated in the pr.

**Comment by YuvHayun** [comment]
The docker pr was merged.

**Comment by shaqnawe** [comment]
> The docker pr was merged.

I have updated the docker image and updated release notes accordingly.

File: Packs/TrendMicroVisionOne/ReleaseNotes/4_1_0.md [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+##### Trend Micro Vision One V3.
+
+- Added new custom script commands:
+  - **trendmicro-visionone-run-custom-script** runs a custom script on an endpoint.
+  - **trendmicro-visionone-get-custom-script-list** fetches a list of available custom scripts in XDR Portal.
+  - **trendmicro-visionone-add-custom-script** adds a custom script to XDR Portal.
+  - **trendmicro-visionone-update-custom-script** updates a custom script in XDR Portal.
+  - **trendmicro-visionone-download-custom-script** downloads a custom script from XDR Portal.
+  - **trendmicro-visionone-delete-custom-script** deletes a custom script from XDR Portal.
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/ReleaseNotes/4_1_0.md:14]:
```suggestion
- Added the following new commands:
  - **trendmicro-visionone-run-custom-script** - runs a custom script on an endpoint.
  - **trendmicro-visionone-get-custom-script-list** - fetches a list of available custom scripts in XDR Portal.
  - **trendmicro-visionone-add-custom-script** - adds a custom script to XDR Portal.
  - **trendmicro-visionone-update-custom-script** - updates a custom script in XDR Portal.
  - **trendmicro-visionone-download-custom-script** - downloads a custom script from XDR Portal.
  - **trendmicro-visionone-delete-custom-script** - deletes a custom script from XDR Portal.
```

**Review by YuvHayun — CHANGES_REQUESTED** [review]
Looks good!
I've left 1 minor comment regarding the RN.
Also, i'll contact you to schedule a quick demo on the new commands you added.

**Comment by shaqnawe** [comment]
> Looks good! I've left 1 minor comment regarding the RN. Also, i'll contact you to schedule a quick demo on the new commands you added.

Sounds good, feel free to reach out on slack if more convenient.

**Comment by YuvHayun** [comment]
Hey @shaqnawe, here are the comments from the demo:
- Update the integration description to contain fewer words, you can add the extra info under the description/readme.md files.
- Add better handling for wrong requests for all the commands - something like "couldn't find a script with the given ID <ID>.".
- Let's convert the file type argument to a predefined list for all commands.
- Let's consider adding a create block object command.
- Let's consider transferring run script command into a polling command with a flag whether to run as polling or not.

Again, thx a lot for your time and let me know if you need any help with the changes.

**Comment by shaqnawe** [comment]
@YuvHayun I have update the error description for delete and download custom script commands but I didn't add polling to run_custom_script command. Additionally the use case of passing multiple values has been requested by our clients so we have decided to keep the current convention of having the users enter a dictionary object. Please let me know if another demo is needed. As of not the PR is ready for merge. Thank you for your assistance!

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3.py [hunk 15]
```diff
@@ -2426,6 +2403,298 @@ def update_status(
...
+    Runs custom script using endpoint (hostname) or agent_guid
+    :type client: ``Client``
+    :param v1_client: pytmv1.Client object used to initialize pytmv1 client.
+    :type args: ``dict``
+    :param args: args object to fetch the argument data.
+    :return: sends data to demisto war room.
+    :rtype: ``dict`
+    """
+    # Required Params
+    block_objects = safe_load_json(args[BLOCK_OBJECTS])
```
Comment (YuvHayun) [inline Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3.py:2419]: Have you decided to skip the block builder?

**Review by YuvHayun — CHANGES_REQUESTED** [review]
Hey @shaqnawe, I have left some comments about things missing from our post demo discussion, let me know what have you decided about it.
No problem about the polling, you can keep it that way.

File: Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3.yml [hunk 16]
```diff
@@ -1190,7 +1190,105 @@ script:
...
+      type: string
+    description: Delete a custom script based on script ID.
+    name: trendmicro-visionone-delete-custom-script
+  - arguments:
+    - description: Name of the custom script. e.g. filename="hello.sh".
+      name: filename
+      required: true
+    - description: The filetype of custom script. e.g. filetype="bash".
+      name: filetype
+      required: true
```
[the remark this answers is not in the document]
Comment (shaqnawe) [reply Packs/TrendMicroVisionOne/Integrations/TrendMicroVisionOneV3/TrendMicroVisionOneV3.yml:1276]: These have been added.

**Comment by shaqnawe** [comment]
@YuvHayun I have update the integration description to make it shorter.

**Comment by shaqnawe** [comment]
@YuvHayun  can we please hold the merge. There's an update for pytmv1 package to address an API issue. I'm working on updating demisto-docker image and will update the dockerimage in yml ASAP. Thank you

**Comment by shaqnawe** [comment]
Created dockerfiles PR in relation to new pytmv1 package update. https://github.com/demisto/dockerfiles/pull/24687

**Comment by YuvHayun** [comment]
Hey @shaqnawe, the docker pr was merged.
Please make sure to pull from master, update docker, and make sure that all steps passed so I can merge the pr.

**Comment by shaqnawe** [comment]
Docker image has been updated to demisto/pytmv1:0.8.2.89444.

**Comment by YuvHayun** [comment]
Great job, thx a lot for your contribution, merging to internal pr.

**Pull request merged**